### PR TITLE
📖  Blogpost announcing CSI snapshotting capability

### PR DESCRIPTION
**Pull request by ashish-amarnath**
Signed-off-by: Ashish Amarnath <ashisham@vmware.com>

### CODE REVIEW

**Comment by skriss** [comment]
Let's check with @jonasrosland on timing of publishing this post vs. the 1.4 release post.

File: site/_posts/2020-05-20-CSI-integration.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
+---
+title: "CSI Volume Snapshotting Capabilities in Velero"
+excerpt: Velero 1.4 now ships with support for CSI snapshotting v1beta1 APIs.
+author_name: Ashish Amarnath
+categories: ['velero','??']
+image: ??
```
Comment (ashish-amarnath) [inline site/_posts/2020-05-20-CSI-integration.md:6]: Not sure what categories mean.
Also, don't know what image to use, and should we use an image.

Comment (skriss) [reply site/_posts/2020-05-20-CSI-integration.md:6]: I don't think we're currently using the categories for anything, presumably it's for a tag cloud/similar.

You get to pick the image! https://unsplash.com/ is a good source for free ones.

File: site/_posts/2020-05-20-CSI-integration.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+```
+
+## Resources
+
+More details about CSI volume snapshotting and its support in Velero may be found in the following links:
+
+- [Kubernetes 1.17 Feature: Kubernetes Volume Snapshot Moves to Beta](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-cis-volume-snapshot-beta/) for more information on the CSI beta snapshot APIs.
+- Prerequisites to use this feature is available [on our website](https://velero.io/docs/csi).
+- [Kubernetes CSI docs](https://kubernetes-csi.github.io/docs/sidecar-containers.html): To understand components in a CSI environment
+- [Velero CSI docs](https://velero.io/docs/csi) for integration and release information.
```
Comment (carlisia) [inline site/_posts/2020-05-20-CSI-integration.md:195]: This link is already on ln 193 above, it shouldn't be repeated here.

File: site/_posts/2020-05-20-CSI-integration.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+1. Set up a CSI environment with a driver supporting the Kubernetes CSI snapshot beta APIs.
+1. Install Velero with CSI snapshotting feature enabled 
+1. Deploy `csi-app`: a pod using a CSI backed PVC that we will backup and restore.
+1. Use Velero to backup and restore the `csi-app`.
+
+## Set up a CSI environment.
+
+As the [Kubernetes CSI Snapshot Beta API](https://kubernetes.io/docs/concepts/storage/volume-snapshots/) is available starting from Kubernetes `1.17`, your will need to be running Kubernetes `1.17` or later.
+
+This post will be using an AKS cluster running Kubernetes `1.17`, with Azure disk CSI driver as an example.
```
Comment (a-mccarthy) [inline site/_posts/2020-05-20-CSI-integration.md:28]: `This post uses an AKS...`

File: site/_posts/2020-05-20-CSI-integration.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+- [`deployment.apps/csi-azuredisk-controller`](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/deploy/csi-azuredisk-controller.yaml)
+- [`daemonset.apps/csi-azuredisk-node`](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/deploy/csi-azuredisk-node.yaml)
+- [`deployment.apps/csi-snapshot-controller`](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/deploy/csi-snapshot-controller.yaml)
+
+## Install Velero with CSI support enabled
+
+The CSI volume snapshot capability is currently, as of version 1.4, a beta feature behind the `EnableCSI` feature flag and is not enabled by default.
+
+Following instructions from our [docs website](https://velero.io/docs/csi/), install Velero with the [velero-plugin-for-csi](https://github.com/vmware-tanzu/velero-plugin-for-csi) and using the Azure Blob Store as our BackupStorageLocation. Please refer to our [velero-plugin-for-microsoft-azure documentation](https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure) for instructions on setting up the BackupStorageLocation.
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply site/_posts/2020-05-20-CSI-integration.md]: I've tried to clarify. PTAL

**Comment by carlisia** [comment]
Sorry @ashish-amarnath I just tuned into this:

image

I'd add an additional text to that summary like: "In this blog post, you'll learn how to setup .... etc". Something like this:

image

File: site/_posts/2020-05-27-CSI-integration.md [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+Using the CSI volume snapshotting features in Velero involves the following steps.
+
+1. Set up a CSI environment with a driver supporting the Kubernetes CSI snapshot beta APIs.
+1. Install Velero with CSI snapshotting feature enabled.
+1. Deploy `csi-app`: a pod using a CSI backed PVC that we will backup and restore.
+1. Use Velero to backup and restore the `csi-app`.
+
+## Set up a CSI environment.
+
+As the [Kubernetes CSI Snapshot Beta API](https://kubernetes.io/docs/concepts/storage/volume-snapshots/) is available starting from Kubernetes `1.17`, your will need to be running Kubernetes `1.17` or later.
```
Comment (jonasrosland) [inline site/_posts/2020-05-27-CSI-integration.md:26]: You can change "your will need to be running" to "you need to run"

File: site/_posts/2020-05-27-CSI-integration.md [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+This post uses an AKS cluster running Kubernetes `1.17`, with Azure disk CSI driver as an example.
+
+Following instructions to install the Azure disk CSI driver from [here](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/docs/install-csi-driver-master.md) run the below command
```
Comment (jonasrosland) [inline site/_posts/2020-05-27-CSI-integration.md:30]: Can change "Following instructions... " to "Follow [these instructions](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/docs/install-csi-driver-master.md) to install the Azure Disk CSI driver, then run the command below:"

File: site/_posts/2020-05-27-CSI-integration.md [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+This script will deploy the following CSI components, CRDs, and necessary RBAC:
+
+- [`deployment.apps/csi-azuredisk-controller`](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/deploy/csi-azuredisk-controller.yaml)
+- [`daemonset.apps/csi-azuredisk-node`](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/deploy/csi-azuredisk-node.yaml)
+- [`deployment.apps/csi-snapshot-controller`](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/deploy/csi-snapshot-controller.yaml)
+
+## Install Velero with CSI support enabled
+
+The CSI volume snapshot capability is currently, as of version 1.4, a beta feature behind the `EnableCSI` feature flag and is not enabled by default.
```
Comment (jonasrosland) [inline site/_posts/2020-05-27-CSI-integration.md:44]: Change "as of version 1.4" to "as of Velero 1.4" to not confuse Velero and K8s versions

File: site/_posts/2020-05-27-CSI-integration.md [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+--plugins velero/velero-plugin-for-microsoft-azure:v1.1.0,velero/velero-plugin-for-csi:v0.1.1 \ 
+--bucket $BLOB_CONTAINER \ 
+--secret-file <PATH_TO_CREDS_FILE>/aks-creds \ 
+--backup-location-config resourceGroup=$AZURE_BACKUP_RESOURCE_GROUP,storageAccount=$AZURE_STORAGE_ACCOUNT_ID,subscriptionId=$AZURE_BACKUP_SUBSCRIPTION_ID \ 
+--snapshot-location-config apiTimeout=5m,resourceGroup=$AZURE_BACKUP_RESOURCE_GROUP,subscriptionId=$AZURE_BACKUP_SUBSCRIPTION_ID \ 
+--image velero/velero:v1.4.0 \ 
+--features=EnableCSI 
+```
+
+NOTE: Please use the latest released version of the container images from the [official Velero dockerhub](https://hub.docker.com/u/velero).
```
Comment (jonasrosland) [inline site/_posts/2020-05-27-CSI-integration.md:62]: Does this mean that the person reading this needs to edit the command above somehow? It's unclear to me.

File: site/_posts/2020-05-27-CSI-integration.md [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+## Deploy a CSI application
```
Comment (jonasrosland) [inline site/_posts/2020-05-27-CSI-integration.md:64]: As you're deploying a stateful app with CSI-enabled backend storage, not a CSI application, can we reword this segment and the text below?

File: site/_posts/2020-05-27-CSI-integration.md [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+Now that the restore has completed and our `csi-nginx` pod is `Running`, confirm that contents of `/mnt/azuredisk/foobar` have been correctly restored.
+
+```bash
+$ kubectl -n csi-app exec -ti csi-nginx bash 
+root@csi-nginx:/# cksum /mnt/azuredisk/foobar 
+2279846381 1726530 /mnt/azuredisk/foobar 
+root@csi-nginx:/# 
+```
+
```
Comment (jonasrosland) [inline site/_posts/2020-05-27-CSI-integration.md:187]: Add a closing segment here, something like "And that's all it takes! Please try out the CSI support in Velero 1.4 and let us know how things go (in Slack, the community meeting, or raise an issue if you find a bug)."

File: site/_posts/2020-05-27-CSI-integration.md [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1Gi
+  storageClassName: disk.csi.azure.com
+---
+```
+
+`kubectl apply 👆 yaml`
```
Comment (nrb) [inline site/_posts/2020-05-27-CSI-integration.md:137]: Do you want the emoji here? I'm honestly not sure what I should run here.

In general, I would aim to write the commands such that someone could copy/paste it.

File: site/_posts/2020-05-27-CSI-integration.md [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+metadata:
+  namespace: csi-app
+  name: csi-nginx
+spec:
+  nodeSelector:
+    kubernetes.io/os: linux
+  containers:
+    - image: nginx
+      name: nginx
+      command: [ "sleep", "1000000" ]
```
Comment (nrb) [inline site/_posts/2020-05-27-CSI-integration.md:113]: Is it your intention to just have the container sleep, or do you want to actually launch nginx?

File: site/_posts/2020-05-27-CSI-integration.md [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+driver: disk.csi.azure.com 
+deletionPolicy: Retain 
+parameters:
+  tags: 'foo=aaa,bar=bbb' 
+```
+
+NOTE: The above `yaml` was sourced from [StorageClass](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/deploy/example/storageclass-azuredisk-csi.yaml) and [VolumeSnapshotClass](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/deploy/example/snapshot/storageclass-azuredisk-snapshot.yaml).
+
+
+Deploy a CSI application: a pod using a CSI backed PVC, in the `csi-app` namespace.
```
Comment (nrb) [inline site/_posts/2020-05-27-CSI-integration.md:93]: Similar to what Jonas has said above, I would reword this to say something like "CSI-backed" application, not "CSI application"

File: site/_posts/2020-05-27-CSI-integration.md [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+```
+
+## Resources
+
+More details about CSI volume snapshotting and its support in Velero may be found in the following links:
+
+- [Kubernetes 1.17 Feature: Kubernetes Volume Snapshot Moves to Beta](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-cis-volume-snapshot-beta/) for more information on the CSI beta snapshot APIs.
+- Prerequisites to use this feature is available [on our website](https://velero.io/docs/csi).
+- [Kubernetes CSI docs](https://kubernetes-csi.github.io/docs/sidecar-containers.html): To understand components in a CSI environment
+- [Velero plugin for CSI](https://github.com/vmware-tanzu/velero-plugin-for-csi) for implementation details of the CSI plugin.
```
Comment (nrb) [inline site/_posts/2020-05-27-CSI-integration.md:195]: Can we change the link text to say `Velero plugin for CSI snapshots`? This will help make it clear what part of the CSI spec we're supporting.

File: site/_posts/2020-05-27-CSI-integration.md [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+  resources:
+    requests:
+      storage: 1Gi
+  storageClassName: disk.csi.azure.com
+---
+```
+
+`kubectl apply 👆 yaml`
+
+Once our pod is running write some data into `/mnt/azuredisk`, where the CSI volume is mounted.
```
Comment (nrb) [inline site/_posts/2020-05-27-CSI-integration.md:139]: You may want to explain why you're writing data into the file instead of relying on an application's data. Is it ultimately so that you can get a consistent checksum value for purposes on demonstration?

File: site/_posts/2020-05-27-CSI-integration.md (hunk 6 above)
Comment (ashish-amarnath) [reply site/_posts/2020-05-27-CSI-integration.md:30]: running that script is part of the instructions. So going to leave this as-is

File: site/_posts/2020-05-27-CSI-integration.md (hunk 8 above)
Comment (ashish-amarnath) [reply site/_posts/2020-05-27-CSI-integration.md:62]: The command by itself can be run as is. 
Removed the "Note:"

File: site/_posts/2020-05-27-CSI-integration.md (hunk 12 above)
Comment (ashish-amarnath) [reply site/_posts/2020-05-27-CSI-integration.md:113]: I want the container to sleep and not want nginx to run.

File: site/_posts/2020-05-27-CSI-integration.md (hunk 15 above)
Comment (ashish-amarnath) [reply site/_posts/2020-05-27-CSI-integration.md:139]: yes exactly. I want to show that the data we restored is exactly the same data we backed up.

**Comment by jonasrosland** [comment]
LGTM!
